### ISSUE TITLE
Fatal error while scanning apple/swift repo

### ISSUE THREAD

**Issue by ptupitsyn** [realm/SwiftLint#4821]
### New Issue Checklist

- [ ] Updated SwiftLint to the latest version
- [ ] I searched for [existing GitHub issues](https://github.com/realm/SwiftLint/issues)

### Describe the bug

SwiftLint docker crashes when I scan https://github.com/apple/swift:

* `git clone https://github.com/apple/swift`
* `cd swift`
* `docker run -it -v `pwd`:`pwd` -w `pwd` ghcr.io/realm/swiftlint:0.50.3`

##### Complete output when running SwiftLint, including the stack trace and command used

```
SourceKittenFramework/File.swift:20: Fatal error: 'try!' expression unexpectedly raised an error: Error Domain=NSCocoaErrorDomain Code=261 "The specified text encoding isn’t applicable."
Current stack trace:
0                                       0x0000559b02f09170 <unavailable> + 30679408
1                                       0x0000559b02bdfa9c <unavailable> + 27363996
2                                       0x0000559b02bdf8b8 <unavailable> + 27363512
3                                       0x0000559b02bde543 <unavailable> + 27358531
4                                       0x0000559b02c3362c <unavailable> + 27706924
5                                       0x0000559b01d4cb03 <unavailable> + 12081923
6                                       0x0000559b02af654c <unavailable> + 26408268
7                                       0x0000559b02aec4c9 <unavailable> + 26367177
8                                       0x0000559b02b0a0fa <unavailable> + 26489082
9                                       0x0000559b02af64c2 <unavailable> + 26408130
10                                      0x0000559b01d4ca00 File.contents.getter + 35
11                                      0x0000559b01df9ba4 <unavailable> + 12790692
12                                      0x0000559b01dfaa6d <unavailable> + 12794477
13                                      0x0000559b01dfaa20 CollectedLinter.styleViolations(using:) + 8
14                                      0x0000559b0257a3fd <unavailable> + 20657149
15                                      0x0000559b02bbc67e <unavailable> + 27219582
16                                      0x0000559b02bbcd6c <unavailable> + 27221356
17                                      0x0000559b02b00845 <unavailable> + 26449989
18                                      0x0000559b02b005f3 <unavailable> + 26449395
19                                      0x0000559b02b0bee2 <unavailable> + 26496738
20   libc.so.6                          0x00007f436e330b43 <unavailable> + 609091
21   libc.so.6                          0x00007f436e3c1b70 clone + 68
```

**Comment by mildm8nnered** [comment]
I think this is actually crashing in SourceKitten - https://github.com/jpsim/SourceKitten/blob/main/Source/SourceKittenFramework/File.swift

**Comment by ptupitsyn** [comment]
From what I can see, there is some file that can't be decoded as utf8. What can we do about that? Any workarounds? It is weird that SwiftLint can't scan Swift itself :)

**Comment by mildm8nnered** [comment]
So I think it's definitely a bug in SourceKitten that will need patching. It does try and defend itself against non-utf8 files, but clearly something is going wrong.

That said, I think there must be something about the setup that is triggering this, as I run SwiftLint against itself all the time (but not on Ubuntu, or via Docker).

**Comment by ptupitsyn** [comment]
> I run SwiftLint against itself all the time

@mildm8nnered can you please try scanning https://github.com/apple/swift ?

I'm on Ubuntu 22, and Docker image is based on Ubuntu 22, so it crashes both in and out of Docker for me.

**Comment by mildm8nnered** [comment]
So  https://github.com/apple/swift is one of the repos that is tested by SwiftLint as part of its CI chain, so I'm pretty sure that that works.

When I run `./tools/oss-check` locally, it passes.

If it's crashing for you with and without Docker, then I guess it's Ubuntu specific, but I don't have an Ubuntu machine to test with :-(

**Comment by ptupitsyn** [comment]
@mildm8nnered thank you for the answer. So it is indeed something about dependencies or the OS setup. Maybe I can try building my own Docker image, based on a different OS.

**Comment by kaypee90** [comment]
I tried linting https://github.com/apple/swift with **SwiftLint version 0.50.3** on my mac and this was the error:

```
Linting '049-swift-typechecker-resolvewitness.swift' (247/16584)
SourceKittenFramework/File.swift:20: Fatal error: 'try!' expression unexpectedly raised an error: Error Domain=NSCocoaErrorDomain Code=261 "The file “033-swift-identifier-isoperatorslow.swift” couldn’t be opened using text encoding Unicode (UTF-8)." UserInfo={NSFilePath=/swift/validation-test/IDE/crashers_fixed/033-swift-identifier-isoperatorslow.swift, NSStringEncoding=4}
PLEASE submit a bug report to https://bugs.llvm.org/ and include the crash backtrace.
[1]    28627 illegal hardware instruction  swiftlint lint
```

Mac Specifications
Os Version: Ventura 13.1
CPU: 2.8 GHz Quad-Core Intel Core i7
Memory: 16 GB 2133 MHz LPDDR3

**Comment by mildm8nnered** [comment]
Hmmm. ok - I see it as well now when I checkout swift and run locally on Mac OS Monterey 12.6.

It does indeed look like that file has some funky characters, and SourceKitten is choking on it because of the first bang here - `_contents = try! String(contentsOfFile: path!, encoding: .utf8)`

What is weird is that `tools/oss-check` doesn't report any problems when linting `swift`, either locally or in CI.

Also, neither the SourceKitten source, nor the file it's choking on have changed for years, so presumably this has been around for a while.

```
% cat ./validation-test/IDE/crashers_fixed/033-swift-identifier-isoperatorslow.swift 
// RUN: %target-swift-ide-test -code-completion -code-completion-token=A -source-filename=%s
??#^A^#
% git log ./validation-test/IDE/crashers_fixed/033-swift-identifier-isoperatorslow.swift
commit 330c2d96e64941c9727c8ce98f43dd05b8783bb6
Author: Ashley Garland <acgarland@apple.com>
Date:   Mon Dec 5 16:22:38 2016 -0800

    Make the lexer UTF-8 RFC 3629 correct re: prefix octets
    
    RFC 2279 states that, in UTF-8:
    "The octet values FE and FF never appear."
    
    RFC 3629 states that, in UTF-8:
    "The octet values C0, C1, F5 to FF never appear."
    
    Generalize the check to advance past invalid starting bytes for
    a UTF-8 sequence to fix a crash in the lexer.
```

**Comment by mildm8nnered** [comment]
So for some reason `oss-check` is only linting a subset of the files:

```
../builds/swiftlint-branch lint --no-cache --enable-all-rules --reporter xcode
Linting Swift files in current working directory
Linting 'Print.swift' (2/376)
Linting 'SwiftOnoneSupport.swift' (1/376)
Linting 'UnsafeRawPointer.swift' (3/376)
Linting 'stdlib/public/core/UnicodeScalarProperties.swift' (4/376)
```

But just running SwiftLint at the top level of `swift` lints a lot more files:

```
% swiftlint lint --no-cache --enable-all-rules --reporter xcode
Linting Swift files in current working directory
Linting 'CxxSetToCollection.swift' (1/16594)
Linting 'CxxStringConversion.swift' (2/16594)
Linting 'ReadAccessor.swift' (3/16594)
Linting 'CreateObjects.swift' (4/16594)
```

so in the `oss-tools` case that file never gets linted and the problem isn't seen there.

**Comment by mildm8nnered** [comment]
ok ... so `oss-check` has special treatment for the `swift` repo - it adds a `.swiftlint.yml`, so that only `stdlib` is scanned, so I guess that's a "workaround". It should be pretty easy to patch this though, as crashing on non-uff-8 files is not great.

```
included: stdlib
```

**Comment by mildm8nnered** [comment]
I have a PR open for SourceKitten that will fix this particular issue - https://github.com/jpsim/SourceKitten/pull/766, once it makes its way into SwiftLint.

However scanning swift will still fail for a number of other reasons :-(

**Comment by mildm8nnered** [comment]
See also https://github.com/realm/SwiftLint/issues/4827 for a different issue that will also cause scanning of apple/swift to fail for multiple files.

**Comment by mildm8nnered** [comment]
And https://github.com/realm/SwiftLint/issues/4829 for another issue with some other files.

**Comment by mildm8nnered** [comment]
and also https://github.com/realm/SwiftLint/issues/4830 affecting one file

**Comment by mildm8nnered** [comment]
and also https://github.com/realm/SwiftLint/issues/4831, for an issue that is not present in `0.50.3`, but is present in main, that is also causing crashes for some files in apple/swift.
